### ISSUE TITLE
SDL can't select an output method other than 'surface' and 'overlay'

### ISSUE THREAD

**Issue by maximilien-noal** [joncampbell123/dosbox-x#212]
For opengl :
Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.
USING AVI+ZMBV
         0 WARN  FPU:FPU80 sizeof(long double) == sizeof(double) so your compiler just makes it an alias. skipping tests. please recompile with proper config.
         0       VGA:Video RAM: 2048KB
         0       SBLASTER:DSP:Reset
         2       INT10:Set Video Mode 3
         2       VGA:Blinking 8
         2       KEYBOARD:INT 15H PS/2 emulation enabled
ISA Plug & Play BIOS enabled
         7       INT10:Set Video Mode 12
         7       VGA:Blinking 0
        31       VGA:VGA refresh rate is now, 59.940
SDL:OPENGL:No support for texturesize of 1024, falling back to surface
SDL:OPENGL:No support for texturesize of 1024, falling back to surface
   2972812       INT10:Set Video Mode 3
   2972812       VGA:Blinking 8
   2972815       BIOS:Keyboard layout fr successfully loaded
DOS keyboard layout loaded with main language code FR for layout fr
   2972815       MISC:UMB assigned region is 0xcb00-0xefff
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file PRN
   2972815       FILES:file open command 2 file CON
   2972815       KEYBOARD:INT 33H emulation enabled
   2973709       VGA:VGA refresh rate is now, 70.087
SDL:OPENGL:No support for texturesize of 1024, falling back to surface
SDL:OPENGL:No support for texturesize of 1024, falling back to surface
   2983970       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2983970       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2984362       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   7971150       SBLASTER:DSP:Reset
--END OF LOG--

For direct3d :

Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.
SDL:Unsupported output device direct3d, switching back to surface
USING AVI+ZMBV
         0 WARN  FPU:FPU80 sizeof(long double) == sizeof(double) so your compiler just makes it an alias. skipping tests. please recompile with proper config.
         0       VGA:Video RAM: 2048KB
         0       SBLASTER:DSP:Reset
         2       INT10:Set Video Mode 3
         2       VGA:Blinking 8
         2       KEYBOARD:INT 15H PS/2 emulation enabled
ISA Plug & Play BIOS enabled
         7       INT10:Set Video Mode 12
         7       VGA:Blinking 0
        31       VGA:VGA refresh rate is now, 59.940
   2969847       INT10:Set Video Mode 3
   2969847       VGA:Blinking 8
   2969850       BIOS:Keyboard layout fr successfully loaded
DOS keyboard layout loaded with main language code FR for layout fr
   2969850       MISC:UMB assigned region is 0xcb00-0xefff
   2969850       FILES:file open command 2 file CON
   2969850       FILES:file open command 2 file CON
   2969850       FILES:file open command 2 file CON
   2969850       FILES:file open command 2 file PRN
   2969850       FILES:file open command 2 file CON
   2969850       KEYBOARD:INT 33H emulation enabled
   2970744       VGA:VGA refresh rate is now, 70.087
   2980990       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2980990       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2981397       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
  13089694       SBLASTER:DSP:Reset
--END OF LOG--

For ddraw :
- DOSBox logo inside DOSBox-X is half-drawn
- log :

Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.
USING AVI+ZMBV
         0 WARN  FPU:FPU80 sizeof(long double) == sizeof(double) so your compiler just makes it an alias. skipping tests. please recompile with proper config.
         0       VGA:Video RAM: 2048KB
         0       SBLASTER:DSP:Reset
         2       INT10:Set Video Mode 3
         2       VGA:Blinking 8
         2       KEYBOARD:INT 15H PS/2 emulation enabled
ISA Plug & Play BIOS enabled
         7       INT10:Set Video Mode 12
         7       VGA:Blinking 0
        31       VGA:VGA refresh rate is now, 59.940
Failed to create ddraw surface, back to normal surface.
   2972812       INT10:Set Video Mode 3
   2972812       VGA:Blinking 8
   2972815       BIOS:Keyboard layout fr successfully loaded
DOS keyboard layout loaded with main language code FR for layout fr
   2972815       MISC:UMB assigned region is 0xcb00-0xefff
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file CON
   2972815       FILES:file open command 2 file PRN
   2972815       FILES:file open command 2 file CON
   2972815       KEYBOARD:INT 33H emulation enabled
   2973709       VGA:VGA refresh rate is now, 70.087
Resetting to WINDIB mode
   2983970       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2983970       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2984362       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
  10398206       SBLASTER:DSP:Reset
--END OF LOG--

For openglnb :
- DOSBox-X crashes as soon as it starts
- log :

Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.

For openglhq :
- DOSBox-X crashes as soon as it starts
- log :

Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.
E_Exit: Could not initialize video: No available video device

For 'overlay' :
- works
- log :

Logging: opened logfile 'log.txt' successfully. All further logging will go to that file.
DOSBox-X version DOSBox-X
         0       MISC:Copyright 2002-2015 enhanced branch by The Great Codeholio, forked from the main project by the DOSBox Team, published under GNU GPL.
         0       GUI:Press Ctrl-F10 to capture/release mouse, Alt-F10 for configuration.
USING AVI+ZMBV
         0 WARN  FPU:FPU80 sizeof(long double) == sizeof(double) so your compiler just makes it an alias. skipping tests. please recompile with proper config.
         0       VGA:Video RAM: 2048KB
         0       SBLASTER:DSP:Reset
         2       INT10:Set Video Mode 3
         2       VGA:Blinking 8
         2       KEYBOARD:INT 15H PS/2 emulation enabled
ISA Plug & Play BIOS enabled
         7       INT10:Set Video Mode 12
         7       VGA:Blinking 0
        31       VGA:VGA refresh rate is now, 59.940
   2966877       INT10:Set Video Mode 3
   2966877       VGA:Blinking 8
   2966880       BIOS:Keyboard layout fr successfully loaded
DOS keyboard layout loaded with main language code FR for layout fr
   2966880       MISC:UMB assigned region is 0xcb00-0xefff
   2966880       FILES:file open command 2 file CON
   2966880       FILES:file open command 2 file CON
   2966880       FILES:file open command 2 file CON
   2966880       FILES:file open command 2 file PRN
   2966880       FILES:file open command 2 file CON
   2966880       KEYBOARD:INT 33H emulation enabled
   2967774       VGA:VGA refresh rate is now, 70.087
   2978020       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2978020       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   2978412       FILES:Special file open command 80 file Z:\AUTOEXEC.BAT
   9058430       SBLASTER:DSP:Reset
--END OF LOG--

---

Configuration :
Using latest release of DOSBox-X.
A test build using the latest sources gives the same results.

Same results on both machines tested.
- The first one uses a NVIDIA GPU (GeForce GTX 950 OC, and an Intel Pentium G CPU
- The second one uses an Intel Core i5 CPU with an Intel HD4000, along side a discrete GeForce 750M GPU (Optimus system)
  Both machines use Windows 10 x64 and a x86 version of DOSBox-X.

**Comment by joncampbell123** [comment]
This fork no longer focuses on Windows development. But there are forks of this code that do focus on Windows development. I'm not familiar enough with DirectX and how DOSBox-X uses it to work on the issue.

**Comment by maximilien-noal** [comment]
Well, that's too sad. :(

So far, DOSBox-X was **by far** my best candidate for running games that demand Windows 9X without using actual (old) hardware. I even have already published solutions which use DOSBox-X with Windows 95 for three different games based both on Macromedia Director 5 and QuickTime 2.x.

As an alternative, would you accept a patch ?

As it seems that older DOSBox-X versions did not have this problem, so I might find what the issue is eventually. But I don't know anything about the SDL, DirectX, or DOSBox-X, so it might take a while. :/

Also could you point me towards forks in which Windows 95 OSR 1 is as stable as with DOSBox-X ?

Thank you very much for DOSBox-X !

**Comment by sikthehedgehog** [comment]
> This fork no longer focuses on Windows development.

OpenGL isn't Windows-specific though (and I get the same "SDL:OPENGL:No support for texturesize of 1024, falling back to surface" error despite being on Ubuntu).

**Comment by Jarvik7** [comment]
Pull requests have been accepted in the past, so I think a patch would be
welcomed.
If you can trace the exact commit that broke what you need fixing it would
be easier as well.

**Comment by joncampbell123** [comment]
I'm OK with integrating patches for Windows support, only that I personally don't bother with Windows development anymore for DOSBox-X.

**Comment by maximilien-noal** [comment]
Does that include Windows 9X emulation ?

@sikthehedgehog :
All I know is that forum posts on VOGONS seem to indicate that 3DFX Win9X games were playable in DOSBox-X, and that now DOSBox-X is very slow with such games, because of the fallback to surface.
I'll have to test old releases first to see if such a tale is possibly true.
I'll see if any can activate OpenGL without falling back to surface.
Then I'll try to see which commit caused it to change.

**Comment by joncampbell123** [comment]
@maximilien-noal No, I mean I don't bother with compiling DOSBox-X to run on Windows. I'm all for Windows support as long as it relates to running Windows as a guest inside DOSBox-X :)

**Comment by maximilien-noal** [comment]
@joncampbell123 :
Great ! :)

As for this issue, I tested every Win32 release of DOSBox-X available on GitHub.

All of them say the same thing :
" SDL:OPENGL:No support for texturesize of 1024, falling back to surface "

Except the very first one, which says that opengl is not a valid option (despite the option being presented)

Also, a lot of them have direct3d as default, yet the logs says this :
"LOG: "direct3d" is not a valid value for variable: output.
It might now be reset to the default value: direct3d "

(But I would be fine with OpenGL)

Sadly it seems a dive in the code is in order, and I know next to nothing about the SDL, OpenGL, and DOSBox-X, apart from toying with the SDL years and years ago.

So if anyone has the knowledge, feel free to investigate this. :)

I'll look into it in my free time but I don't think I'll get anywhere.

**Comment by maximilien-noal** [comment]
Ok, first I disabled this check :

`if (texsize >= sdl.opengl.max_texsize) {
			LOG_MSG("SDL:OPENGL:No support for texturesize of %d, falling back to surface", texsize);
			goto dosurface;
		}
`

Because it seems outdated and therefore useless nowadays :

http://stackoverflow.com/questions/19572659/is-1024x1024-a-widely-supported-opengl-maximum-texture-size-on-the-desktop

> According to this SO answer "what is the max size of the texture iphone can support?" the iPhone 3g can support 1024x1024 texture sizes. That phone was released in 2008, which leads me to believe that most desktop would be able to support the same 5 years ago.

The GL_MAX_TEXTURE_SIZE is defined by the implementation. Wild Fire Games has also been kind enough to offer their OpenGL capabilities report which has max texture sizes from their data set. As of writing this, only 18 out of 33085 (0.05%) users have reported to not have the ability to render a GL_MAX_TEXTURE_SIZE 1024 or more.

Besides, the question was about smartphones, and it was from 2013.

And it wasn't working anyway since the SDL always reported 0. I could not find documentation on it.

Then I ran the 3dfx "Grand Bleu" demo (downloaded from here : http://www.falconfly.de/artwork.htm ) in DOSBox-X for 5 hours in Windows 95 OSR1.

It worked fine, and even better with memsize=32 (the demo demands at least 32 MB of RAM)

Although sometimes DOSBox-X crashes at startup (it never did that with output=surface) on a segmentation fault. I haven't looked into it yet.

Anyway here is the patch :
Subject: [PATCH] disabled max_texsize opengl check (obsolete)

http://stackoverflow.com/questions/19572659/is-1024x1024-a-widely-supported-opengl-maximum-texture-size-on-the-desktop
-> only 0.13 % of smartphones in 2013 did not support 1024*1024
---
 src/gui/sdlmain.cpp | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/src/gui/sdlmain.cpp b/src/gui/sdlmain.cpp
index 04ead86..a0c3720 100644
--- a/src/gui/sdlmain.cpp
+++ b/src/gui/sdlmain.cpp
@@ -1041,11 +1041,6 @@ dosurface:
 		sdl.opengl.framebuf=0;
 		//if (!(flags&GFX_CAN_32) || (flags & GFX_RGBONLY)) goto dosurface;
 		// SDLScreen_Reset();
-		int texsize=2 << int_log2(width > height ? width : height);
-		if (texsize>sdl.opengl.max_texsize) {
-			LOG_MSG("SDL:OPENGL:No support for texturesize of %d, falling back to surface",texsize);
-			goto dosurface;
-		}
 		SDL_GL_SetAttribute( SDL_GL_DOUBLEBUFFER, 1 );
 #if SDL_VERSION_ATLEAST(1, 2, 11)
 		Section_prop * sec=static_cast<Section_prop *>(control->GetSection("vsync"));
-- 

No pull request yet, since there is that semi-random segmentation fault problem at startup.

For the segmentation-fault that happens sometimes (not always, which make it hard to debug), I wish to leave it to more capable hands.

**Comment by masaykh** [comment]
So did solution found for this problem on windows?
SDL:OPENGL:No support for texturesize of 1024, falling back to surface
currently only surface mode available...

**Comment by maximilien-noal** [comment]
It works when you disable the outdated/useless code that checked the maximum supported texture size (which always gets 0 from the SDL and therefore goes back to surface mode),
 
But I still don't have a clue why it makes DOSBox crash at startup sometimes, while other times it's doing fine. Same game, same config, different outcome = Why ?!

It really irks me, but sadly DOSBox's code is way out my league.

**Comment by joncampbell123** [comment]
I haven't been able to test that code, I don't develop on Windows.

@maximilien-noal I don't blame you, DOSBox's source code to me comes off as what happens when you let 10 years of development + hacks pile up :) I have taken efforts to clean it up and organize it, but there's more mess than I can clean up.

If you have a debugger handy, see where it crashes and get a stack trace. This should be possible in Visual Studio 2015 or if you can get Cygwin or MinGW and the GDB debugger to catch it.

**Comment by maximilien-noal** [comment]
It works when you disable the outdated/useless code that checked the maximum supported texture size (which always gets 0 from the SDL and therefore goes back to surface mode),
 
But I still don't have a clue why it makes DOSBox crash at startup sometimes, while other times it's doing fine. Same game, same config, different outcome = Why ?!

It really irks me, but sadly DOSBox's code is way out my league.

**Comment by maximilien-noal** [comment]
On Linux, it is the same deal : each time output=opengl is selected, it goes back to surface. So I think you could debug it on Linux ?

In fact, this very bug exists in vanilla DOSBox v0.74.

No idea if you will get the random startup crashes once you remove the max_texsize test, however.

Thanks for the idea about stacktraces, I'll use it once the irking will be too bothersome to ignore. ;-)

**Comment by masaykh** [comment]
I work with java so no C\C++ toolchain available , can someone make fixed .exe ?
and btw - can dosbox-x be compiled with cygwin\ mingw ?

**Comment by joncampbell123** [comment]
I would assume so, never tried it. Anyone here confirm DOSBox compiles with Windows + Cygwin?

**Comment by box-tortoise** [comment]
I can work on that. :)

**Comment by Silanda** [comment]
FWIW, debugging with Visual Studio, the intermittent crash occurs when the function "void GFX_LogSDLState(void)", on line 694, is called in sdlmain.cpp. However, I'm not a programmer, and Dosbox's code is pretty daunting, so I honestly don't know where to begin with tracking down the problem.

**Comment by aybe** [comment]
TL;DR there's a bunch of code that simply doesn't work anymore, i.e. D3D. I'll see whether I'll disable the menu items or delete them to make it less confusing, rule of thumb would be : what's in the menus are what's available (fixing the dosbox.conf will be longer)

**Comment by martinlindhe** [comment]
I think this was fixed by https://github.com/joncampbell123/dosbox-x/pull/254

**Comment by aybe** [comment]
Right but OpenGL (at least here) has screen corruption happening randomly, requiring a switch to fullscreen or a scaler change:

![2017-09-14 10_03_16-dosbox-x dosbox-x max 100 cyc_ms dosbox](https://user-images.githubusercontent.com/1537591/30418694-27fe3f70-9934-11e7-90be-0481dd138e92.png)

![2017-09-14 10_03_30-dosbox-x dosbox-x max 100 cyc_ms dosbox](https://user-images.githubusercontent.com/1537591/30418696-2c471f16-9934-11e7-92b6-0f612e61617a.png)

i.e. should be looked at in detail to track the cause of this

**Comment by martinlindhe** [comment]
@aybe Okay. I have not experienced random corruption myself. But I have not even tried to run dosbox-x on Windows.
Can you describe a little more what is going on in the second screenshot (the first looks good, right?)
Is the second screenshot supposed to show the same content as the first one?

I have only tested the #254 in macOS and Linux (using nvidia binary driver).

I have a nvidia graphics card, so I will only be able to test this out on Windows 10 using nvidia drivers.
What OpenGL drivers are you using on Windows, nvidia, amd or intel?

Also, are you running Windows 10 or other version?

**Comment by aybe** [comment]
I set to OpenGL, go fullscreen, go window again, OGL rendering is broken and have to switch to surface again for it to work.

Win10, AMD, it is very possible that can be due to the AMD GPU especially drivers.

**Comment by joncampbell123** [comment]
I would like to know if this is still a bug.

On Linux, I've been able to use "Surface" and "OpenGL" outputs.

On Windows, I've been able to use Surface, Direct3D, OpenGL, etc.

If not, I will close this.

**Comment by maximilien-noal** [comment]
It is indeed fixed. I was able to run DOSBox-X v12-24-2017 on Windows 10 FCU with OpenGL enabled several times without any crash.

Thank you so much !
